zerver/webhooks/sonarqube/view.py

```diff
@@ -0,0 +1,104 @@
+# Webhooks for external integrations.
+
+from typing import Any, Dict, List, Mapping
+
+from django.http import HttpRequest, HttpResponse
+
+from zerver.decorator import webhook_view
+from zerver.lib.request import REQ, has_request_variables
+from zerver.lib.response import json_success
+from zerver.lib.webhooks.common import check_send_webhook_message
+from zerver.models import UserProfile
+
+inverse_operators = {
+    "WORSE_THAN": "should be better or equal to",
+    "GREATER_THAN": "should be less than or equal to",
+    "LESS_THAN": "should be greater than or equal to",
+}
+
+templates = {
+    "default": "* {}: **{}** {} {} {}.",
+    "no_value": "* {}: **{}**.",
+}
+
+
+def parse_metric_name(metric_name: str) -> str:
+    return " ".join(metric_name.split("_"))
+
+
+def parse_condition(condition: Mapping[str, Any]) -> str:
+    metric = condition["metric"]
+
+    metric_name = parse_metric_name(metric)
+    operator = condition["operator"]
+    operator = inverse_operators.get(operator, operator)
+    value = condition.get("value", "no value")
+    status = condition["status"].lower()
+    threshold = condition["errorThreshold"]
+
+    if value == "no value":
+        return templates["no_value"].format(metric_name, status)
+
+    template = templates["default"]
+
+    return template.format(metric_name, status, value, operator, threshold)
+
+
+def parse_conditions(conditions: List[Mapping[str, Any]]) -> str:
+    return "\n".join(
+        [
+            parse_condition(condition)
+            for condition in conditions
+            if condition["status"].lower() != "ok" and condition["status"].lower() != "no_value"
+        ]
+    )
+
+
+def render_body(payload: Mapping[str, Any]) -> str:
+    project_name = payload["project"]["name"]
+    project_url = payload["project"]["url"]
+    quality_gate_status = payload["qualityGate"]["status"].lower()
+    if quality_gate_status == "ok":
+        quality_gate_status = "success"
+    else:
+        quality_gate_status = "error"
+    branch = None
+    if "branch" in payload.keys():
+        branch = payload["branch"].get("name", None)
+    conditions = payload["qualityGate"]["conditions"]
+    conditions = parse_conditions(conditions)
+    if branch:
+        if not conditions:
+            return "Project [{}]({}) analysis of branch {} resulted in {}.".format(
+                project_name, project_url, branch, quality_gate_status
+            )
+        msg = "Project [{}]({}) analysis of branch {} resulted in {}:\n".format(
+            project_name, project_url, branch, quality_gate_status
```

Comment:
Since we are using these message formatting templates multiple times, I would recommend moving these to their constant at the top of the file as well. E.g. names "MESSAGE_WITH_CONDITIONS", "MESSAGE_WITHOUT_CONDITIONS".

Reply:
Sure